src/spec-lang/tc/builtins.ts

```diff
@@ -1,143 +1,178 @@
-import { DataLocation, FunctionStateMutability, FunctionVisibility } from "solc-typed-ast";
 import {
-    SAddressType,
-    SBoolType,
-    SBuiltinStructType,
-    SBytes,
-    SFixedBytes,
-    SFunctionType,
-    SIntType,
-    SPointer,
-    SType
-} from "../ast";
+    DataLocation,
+    FunctionStateMutability,
+    FunctionVisibility,
+    AddressType,
+    BoolType,
+    BytesType,
+    FixedBytesType,
+    FunctionType,
+    IntType,
+    PointerType,
+    TypeNode,
+    Range,
+    BuiltinType
+} from "solc-typed-ast";
 
-export const BuiltinSymbols = new Map<string, SType>([
-    ["abi", new SBuiltinStructType("abi", new Map())],
+export class BuiltinStructType extends BuiltinType {
+    public readonly members: Map<string, TypeNode>;
+
+    constructor(name: string, members: Map<string, TypeNode>, src?: Range) {
+        super(name, src);
+        this.members = members;
+    }
+
+    pp(): string {
+        return `builtin_struct ${this.name}`;
+    }
+
+    getFields(): any[] {
+        return [this.name];
+    }
+}
+
+export const BuiltinSymbols = new Map<string, TypeNode>([
```

Comment:
Missing `bytes.concat` here I guess (introduced in [Solidity 0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4))

Reply:
`bytes.concat` is difficult to add because it supports a variable number of arguments. Supporting variable-number-of-arguments builtins requires re-architecting some stuff here. Since this is larger, lets do it as a separate PR